src/Adapter/Discount/Update/DiscountDuplicator.php

```diff
@@ -0,0 +1,331 @@
+
+        foreach ($shops as $shopId) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_shop')
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'id_shop' => ':id_shop',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('id_shop', $shopId)
+                ->executeQuery();
+        }
+    }
+
+    /**
+     * Duplicates product rules associated to a discount
+     */
+    private function duplicateProductRules(DiscountId $oldDiscountId, DiscountId $newDiscountId): void
+    {
+        // First, we need to copy the rule groups (with their new IDs)
+        $ruleGroupsIds = [];
+        $qb = $this->connection->createQueryBuilder();
+        $ruleGroups = $qb
+            ->select('*')
+            ->from($this->dbPrefix . 'cart_rule_product_rule_group')
+            ->where('id_cart_rule = :id_discount')
+            ->setParameter('id_discount', $oldDiscountId->getValue())
+            ->executeQuery()
+            ->fetchAllAssociative();
+
+        foreach ($ruleGroups as $ruleGroup) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_product_rule_group')
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'quantity' => ':quantity',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('quantity', $ruleGroup['quantity'])
+                ->setParameter('type', $ruleGroup['type'])
+                ->executeQuery();
+            $ruleGroupsIds[$ruleGroup['id_product_rule_group']] = $this->connection->lastInsertId();
+        }
+
+        // Then, we can copy the rules, linking them to the new group IDs
+        $rulesIds = [];
+        $qb = $this->connection->createQueryBuilder();
+        $rules = $qb
+            ->select('*')
+            ->from($this->dbPrefix . 'cart_rule_product_rule')
+            ->where('id_product_rule_group IN (:id_rule_groups)')
+            ->setParameter('id_rule_groups', array_keys($ruleGroupsIds), ArrayParameterType::INTEGER)
+            ->executeQuery()
+            ->fetchAllAssociative();
+
+        foreach ($rules as $rule) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_product_rule')
+                ->values([
+                    'id_product_rule_group' => ':id_rule_group',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_rule_group', $ruleGroupsIds[$rule['id_product_rule_group']])
+                ->setParameter('type', $rule['type'])
+                ->executeQuery();
+            $rulesIds[$rule['id_product_rule']] = $this->connection->lastInsertId();
```

Comment:
Same suggestion regarding the rule Ids, seems like you are mixing old and new values in the same array